include/elements/2d/quadrilateral_bspline_element.h

```diff
@@ -0,0 +1,245 @@
+
+  //! Compute Jacobian local
+  //! \param[in] xi given local coordinates
+  //! \param[in] nodal_coordinates Coordinates of nodes forming the cell
+  //! \param[in] particle_size Particle size
+  //! \param[in] deformation_gradient Deformation gradient
+  //! \retval jacobian Jacobian matrix
+  Eigen::Matrix<double, Tdim, Tdim> jacobian_local(
+      const VectorDim& xi, const Eigen::MatrixXd& nodal_coordinates,
+      const VectorDim& particle_size,
+      const VectorDim& deformation_gradient) const override;
+
+  //! Evaluate the B matrix at given local coordinates for a real cell
+  //! \param[in] xi given local coordinates
+  //! \param[in] nodal_coordinates Coordinates of nodes forming the cell
+  //! \param[in] particle_size Particle size
+  //! \param[in] deformation_gradient Deformation gradient
+  //! \retval bmatrix B matrix
+  std::vector<Eigen::MatrixXd> bmatrix(
+      const VectorDim& xi, const Eigen::MatrixXd& nodal_coordinates,
+      const VectorDim& particle_size,
+      const VectorDim& deformation_gradient) const override;
+
+  //! Return the type of shape function
+  mpm::ShapefnType shapefn_type() const override {
+    return mpm::ShapefnType::BSPLINE;
+  }
+
+  //! Return number of shape functions
+  unsigned nfunctions() const override { return nconnectivity_; }
+
+  //! Return if natural coordinates can be evaluates
+  bool isvalid_natural_coordinates_analytical() const override { return false; }
+
+  //! Compute Natural coordinates of a point (analytical)
+  //! \param[in] nodal_coordinates Coordinates of nodes forming the cell
+  //! \param[in] point Location of the point in cell
+  //! \retval xi Return the local coordinates
+  VectorDim natural_coordinates_analytical(
+      const VectorDim& point,
+      const Eigen::MatrixXd& nodal_coordinates) const override;
+
+  //! Assign nodal connectivity property for bspline elements
+  //! \param[in] nodal_coordinates Coordinates of nodes forming the cell
+  //! \param[in] nodal_properties Vector determining node type for each
+  //! dimension
+  void initialise_bspline_connectivity_properties(
+      const Eigen::MatrixXd& nodal_coordinates,
+      const std::vector<std::vector<unsigned>>& nodal_properties) override;
+
+  //! Return the degree of shape function
+  mpm::ElementDegree degree() const override {
+    return mpm::ElementDegree::Quadratic;
+  };
+
+ private:
+  //! Compute B-Spline Basis Function using the recursive De Boor's algorithm
+  //! for single direction
+  //! \param[in] point_coord point coordinate in one direction
+  //! \param[in] nodal_coord nodal coordinate in one direction
+  //! dimension
+  //! \param[in] node_type Node type associated with direction
+  //! \param[in] poly_order Polynomial degree
+  //! \param[in] index Index associated to local BSplineKnotVector
+  double kernel(double point_coord, double nodal_coord, unsigned node_type,
+                unsigned poly_order, unsigned index = 0) const;
```

Comment:
Is "kernel" a typical name for this? I've not seen "kernel" use with Cox De Boor (but I also don't have any experience with B-splines in the wild)

Reply:
Good point! I took it from my experience with SPH kernels. Which basically means the weighting function according to distance. I think it's not a typical name in FEM variant methods.

Reply:
@jgiven100 I ended up looking in Wikipedia for the origin of this function in SPH. It is somehow the same but might be a bit different. See https://en.wikipedia.org/wiki/Kernel_smoother

What do you think? I can change if needed. Any suggestions?

Reply:
It does look like its a bit different... But since I don't have any good suggestions for a name change, its probably ok as "kernel". 

Also, 
```
  //! Compute B-Spline Basis Function using the recursive De Boor's algorithm
```
is pretty clear and easy to find with `ctl+F`